lib/evm/opFns.ts

```diff
@@ -1,45 +1,57 @@
-const Buffer = require('safe-buffer').Buffer
-const utils = require('ethereumjs-util')
-const BN = utils.BN
-const { ERROR, VmError } = require('../exceptions')
+import BN = require('bn.js')
+import * as utils from 'ethereumjs-util'
+import { ERROR, VmError } from '../exceptions'
+import { RunState } from './loop'
+
 const MASK_160 = new BN(1).shln(160).subn(1)
 
 // Find Ceil(`this` / `num`)
-BN.prototype.divCeil = function divCeil (num) {
-  var dm = this.divmod(num)
+function divCeil (a: BN, b: BN) {
+  const div = a.div(b)
+  const mod = a.mod(b)
 
   // Fast case - exact division
-  if (dm.mod.isZero()) return dm.div
+  if (mod.isZero()) return div
 
   // Round up
-  return dm.div.negative !== 0 ? dm.div.isubn(1) : dm.div.iaddn(1)
+  return div.isNeg() ? div.isubn(1) : div.iaddn(1)
 }
 
-function addressToBuffer (address) {
+function addressToBuffer (address: BN) {
   return address.and(MASK_160).toArrayLike(Buffer, 'be', 20)
 }
 
+export interface SyncOpHandler {
+  (runState: RunState): void
+}
+
+export interface AsyncOpHandler {
+  (runState: RunState): Promise<void>
+}
+
+export type OpHandler = SyncOpHandler | AsyncOpHandler
+
 // the opcode functions
-module.exports = {
-  STOP: function (runState) {
-    runState.stopped = true
+export const handlers: {[k: string]: OpHandler} = {
```

Comment:
Is this mainly to closely stick to the current structure of the module or does this have other benefits or - more broadly - implications?

Related question: is making opcodes sub-classable or extendable or similar a separate step? Have you this on the plan for this release/change round?

Reply:
Well, no major benefit, I only typed what was there, didn't make structural modifications. I think it definitely makes sense to revisit this and maybe take inspiration from Chris's comments to improve the handlers.

>  is making opcodes sub-classable or extendable or similar a separate step?

Once users can customize `Loop` and use their customized version to run transactions, they can override `getOpHandler` and replace the implementation for any of the opcodes. I imagine this should be a good first step. I'm open however to ideas on how to make this even further extensible.

Reply:
Ok, thanks for the explanation.